### ISSUE TITLE
adblock: Support cosmetic filtering (element hiding) and scriptlets

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#6480]
Splitting this off from #5754 since it's clearly the most important part of it, and also requires some more text.

The underlying Brave adblocking library seems to already support both [element hiding](https://adblockplus.org/filter-cheatsheet#elementhiding) and [Scriptlets](https://github.com/AdguardTeam/Scriptlets), but qutebrowser doesn't support them so far. Not sure if they're supported in the python-adblock library or if some more work is required there.

Original discussion here: https://github.com/qutebrowser/qutebrowser/pull/5317#issuecomment-612554553.
Related: #6460

Scriptlets seem to be required to properly block YouTube ads after the latest changes, see e.g. [this post by the easylist maintainer](https://forums.lanik.us/viewtopic.php?p=160013&sid=f616161429ceb64e4cdec0a0edb27772#p160013):

> If you don't like youtube ads, switching browsers (Chrome, Firefox, Brave) will be the only option. Basic network blocking won't be enough to stop youtube ads unfortantly

and [this post](https://forums.lanik.us/viewtopic.php?p=160029#p160029):

> Thanks, implementing override-property-read, json-prune, and using those extra two rules did the job!

Some more resources:

- [YouTube ads in Safari: you see them now, will you see them in the future?](https://adguard.com/en/blog/youtube-ads-in-safari-explained.html)
- [225861 – Content Blocker: add a new action type - "scriptlet"](https://bugs.webkit.org/show_bug.cgi?id=225861)
- [Allow Snippet use in Easylist / Easyprivacy (#224) · Issues · eyeo / adblockplus / adblockpluscore · GitLab](https://gitlab.com/eyeo/adblockplus/adblockpluscore/-/issues/224)
- [Resources Library · gorhill/uBlock Wiki](https://github.com/gorhill/uBlock/wiki/Resources-Library)
- [uBlock/scriptlets.js at master · gorhill/uBlock](https://github.com/gorhill/uBlock/blob/master/assets/resources/scriptlets.js)
- [WIP: added scriptlet support (a1a79b88) · Commits · Allan Sandfeld Jensen / Falkon · GitLab](https://invent.kde.org/carewolf/falkon/commit/a1a79b882b70cef77ed37e4086a4b862bc648d52)

cc @ArniDagur

**Comment by ArniDagur** [comment]
> The underlying Brave adblocking library seems to already support both element hiding and Scriptlets, but qutebrowser doesn't support them so far. Not sure if they're supported in the python-adblock library or if some more work is required there.

There are some things that are supported in `python-adblock`, and some that are missing, e.g. resources. It shouldn't be too much effort to support everything.

I don't have as much time for new open source contributions as I used to, but I will add any functionality to `python-adblock` that is needed. Additionally, if someone can provide a guide on how to

* Get all of the classes and ids of the web page's elements from qutebrowser.
* Inject CSS and JavaScript into a web page from qutebrowser.

then I can take a stab at cosmetic blocking.

**Comment by sverona** [comment]
> Additionally, if someone can provide a guide on how to
> 
> * Get all of the classes and ids of the web page's elements from qutebrowser.
> * Inject CSS and JavaScript into a web page from qutebrowser.
> 
> then I can take a stab at cosmetic blocking.

I couldn't find an entry in the QB API for doing either of these, so you may have to go through WebEngine. [This looks like the primary resource.](https://doc.qt.io/qt-5/qtwebengine-overview.html#script-injection) However:

> Qt WebEngine does not allow direct access to the document object model (DOM) of a page. However, the DOM can be inspected and adapted by injecting scripts.

So my guess is you'd need to run some JS through a QWebChannel to pull classes and IDs. [Here is the best reference I could find describing how to do this.](https://stackoverflow.com/questions/47252632/how-to-pass-info-from-js-to-python-using-qwebchannel)

[Here is a reference that describes how to inject JS into a page.](https://doc.qt.io/qt-5/qtwebengine-webenginewidgets-contentmanipulation-example.html)

I don't know enough about the internals of WebEngine or Brave's adblock to say how this will impact performance.

**Comment by The-Compiler** [comment]
Hopefully, QtWebChannel won't be needed, and any implementation using it will be heavily scrutinized due to its security impact (if you set up a QtWebChannel in the same context web pages run, it follows that web pages can use that QtWebChannel as well, and call things which might be qutebrowser-internal).

There are two ways to run JS with QtWebEngine:

- [QWebEnginePage::runJavaScript](https://doc.qt.io/qt-5/qwebenginepage.html#runJavaScript), exposed in qutebrowser's tab API [via `tab.run_js_async()`](https://github.com/qutebrowser/qutebrowser/blob/v2.2.3/qutebrowser/browser/browsertab.py#L1174-L1191). It takes a callback which gets whatever was returned from JS, so this should be enough to find the elements. It's also what's used to implement hints and other features already, via the [`tab.elements` API](https://github.com/qutebrowser/qutebrowser/blob/v2.2.3/qutebrowser/browser/browsertab.py#L704-L762).
- [`QWebEngineScript`](https://doc.qt.io/qt-5/qwebenginescript.html) which is more like a Greasemonkey script (and in fact supports a sub-set of Greasemonkey comments), i.e. you "install" a script and QtWebEngine/Chromium will take care of automatically injecting it on every page load (or all page loads matching a pattern). This isn't exposed via the tab API so far (and also only available with QtWebEngine), but it's used internally by `webenginetab.py` in [`_WebEngineScripts`](https://github.com/qutebrowser/qutebrowser/blob/v2.2.3/qutebrowser/browser/webengine/webenginetab.py#L1003-L1200) for various built-in functionality, Greasemonkey support, etc. etc.

Unfortunately, there's no way to inject style sheets - it all needs to be done via JS. See the [Greasemonkey wrapper](https://github.com/qutebrowser/qutebrowser/blob/v2.2.3/qutebrowser/javascript/greasemonkey_wrapper.js#L96-L108) for a simple example how to do this, and [stylesheet.js](https://github.com/qutebrowser/qutebrowser/blob/v2.2.3/qutebrowser/javascript/stylesheet.js) for a more complex one supporting live updates (used for `content.user_stylesheets`).

So, yeah, some of the infrastructure probably is already in place (after, all, there's [jhide](https://gitlab.com/jgkamat/jhide) using the Greasemonkey support for cosmetic adblock filters already), but getting this all to work properly is probably still not too trivial.

@ArniDagur can you elaborate a bit more about what kind of APIs you'd need, and how those cosmetic rules and the adblock rust API work? Why do you need a list of all classes/ids, for example?

**Comment by ArniDagur** [comment]
> @ArniDagur can you elaborate a bit more about what kind of APIs you'd need, and how those cosmetic rules and the adblock rust API work? Why do you need a list of all classes/ids, for example?

The best resource to understand this is the following GitHub issue: https://github.com/brave/adblock-rust/issues/152. 

> But roughly, first you'd want to call `url_cosmetic_resources`, of which three fields are relevant: `hide_selectors`, `style_selectors`, and `exceptions`. Then you'll want to gather all of the CSS classes and ids that occur on the page, and call `hidden_class_id_selectors`, being sure to pass the `exceptions` as the third argument.
> 
>  * `hide_selectors` is a list of CSS selectors to which `display: none !important` should be applied.
>  * The result from `hidden_class_id_selectors` is another list of simple CSS selectors to which `display: none !important` should be applied.
>  * `style_selectors` maps CSS selectors on the page to corresponding rules other than `display: none !important`.
>

**Comment by The-Compiler** [comment]
Thanks! That together with https://github.com/brave/adblock-rust/issues/152#issuecomment-771046487 and the docs added in https://github.com/brave/adblock-rust/commit/445b633d94e4139cce94a154acf530f9a449f945 gives me rough idea on how that'd work - and it's much more complex than I would've thought...

Especially reacting to new elements appearing on a page dynamically (which is somewhat common nowadays) could be tricky. They say:

> we only inject rules based on classes and ids that actually appear on the page (in practice, we use a MutationObserver to identify those elements)

but something like that would indeed require a `QWebChannel` - something I'm still quite wary about, given the security implications of letting JS running in the page's context talk back to Python.

From a quick look how [Falkon](https://www.falkon.org/) solves this, it looks like they do indeed just inject all rules into every page...

https://github.com/KDE/falkon/blob/8abf9d5cf0e682e9c16fa8604d41299d8327f9ad/src/lib/adblock/adblockplugin.cpp#L80-L89

https://github.com/KDE/falkon/blob/8abf9d5cf0e682e9c16fa8604d41299d8327f9ad/src/lib/adblock/adblockmatcher.cpp#L179-L206

**Comment by The-Compiler** [comment]
As an aside, for people looking for a workaround for YouTube: Someone on Reddit [recently wrote a Greasemonkey script](https://old.reddit.com/r/qutebrowser/comments/ntl2ko/easy_youtube_adblocker_greasemonkey_script/h15yaqc?utm_source=share&utm_medium=web2x&context=3) to do so. I haven't tested it myself yet. Another approach is to use an external player like `mpv` instead, see 10. in the [FAQ](https://github.com/qutebrowser/qutebrowser/blob/master/doc/faq.asciidoc).

**Comment by sverona** [comment]
> As an aside, for people looking for a workaround for YouTube: Someone on Reddit [recently wrote a Greasemonkey script](https://old.reddit.com/r/qutebrowser/comments/ntl2ko/easy_youtube_adblocker_greasemonkey_script/h15yaqc?utm_source=share&utm_medium=web2x&context=3) to do so. I haven't tested it myself yet. Another approach is to use an external player like `mpv` instead, see 10. in the [FAQ](https://github.com/qutebrowser/qutebrowser/blob/master/doc/faq.asciidoc).

The Greasemonkey script is a great stopgap.

Here is a mirror in case something happens to the Reddit post. I have this in `~/.local/share/qutebrowser/greasemonkey/youtube-adblock.user.js`.

```js
// ==UserScript==
// @name         Auto Skip YouTube Ads 
// @version      1.0.1
// @description  Speed up and skip YouTube ads automatically 
// @author       jso8910
// @match        *://*.youtube.com/*
// @exclude      *://*.youtube.com/subscribe_embed?*
// ==/UserScript==
let main = new MutationObserver(() => {
    let ad = [...document.querySelectorAll('.ad-showing')][0];
    if (ad) {
        let btn = document.querySelector('.videoAdUiSkipButton,.ytp-ad-skip-button')
        if (btn) {
            btn.click()
        }
    }
})

main.observe(document.querySelector('.videoAdUiSkipButton,.ytp-ad-skip-button'), {attributes: true, characterData: true, childList: true})
```

**Comment by crocket** [comment]
This issue should be prioritized. Unless qutebrowser can be configured to automatically replace www.youtube.com with yewtu.be, qutebrowser is painful on youtube.

**Comment by The-Compiler** [comment]
@crocket There are various workarounds for Youtube specifically, see e.g. the comment right above yours.

**Comment by crocket** [comment]
My workaround is to make qutebrowser delegate URLs to a URL handler userscript that turns www.youtube.com and youtube.com into yewtu.be

**Comment by metov** [comment]
@The-Compiler I think this feature would be amazing and I really hope it can be added to qutebrowser soon. Qutebrowser is already a great piece of software and is a far superior experience to even chromium with vimium and the like - just native support for keyboard navigation alone already puts it miles ahead of the competition. Unfortunately the web these days is basically unusable without sophisticated request blocking and element hiding - between all the cookie banners, ads, analytics scripts, javascript bloat it's very difficult to actually use most websites "as intended". This feature, if implemented properly, would allow me to finally switch to qb as my main browser for good. And I suspect I'm not the only one.

I'd love to help out with the development of this. However the qb codebase, currently at 75k lines, isn't quite as straightforward to just dive in (and in this case, the real problem is probably the complexity of Brave's codebase rather than qb). Is there any chance we could break this issue down into smaller chunks, that are more self-contained and realistic for a new contributor to try and attack without having to understand the entire codebase?

For example, are there any limited-scope MVP's or proof of concepts you can see that people could contribute and move the effort forward? If I wanted to help with this, where would I start? I've read through the issue and I have a general idea, but it's not really clear to me what the steps and priorities are.

I realize that this breaking down into subtasks will require work on your part as well, probably quite substantial work in fact. But it would also enable other people to contribute and so might be less work than just implementing the whole thing yourself. So could we maybe have something like a roadmap for how we would be getting from here to having fully functional element hiding in qb?

**Comment by The-Compiler** [comment]
Perhaps relevant for people looking for a YouTube workaround: [Updated Auto Skip Youtube Ads : qutebrowser](https://www.reddit.com/r/qutebrowser/comments/qe4mwi/updated_auto_skip_youtube_ads/)

**Comment by herrsimon** [comment]
Maybe this comment is very naive, but how about https://github.com/ghostery/adblocker? Apparently a pure javascript library (so could be run as a simple background script) and claimed to support '99% of all filters from the Easylist and uBlock Origin project'. At least as an interim solution, it seems to me that this should fit the bill, or is there something I am overlooking?

**Comment by The-Compiler** [comment]
@herrsimon We don't have anything like extension background scripts. Maybe with some effort it could be made into a greasemonkey script somehow, but I'd rather put effort into properly integrating the existing adblocker library rather than integrating a second one.

**Comment by herrsimon** [comment]
I completely agree that properly integrating the existing library is a better solution and was thinking of a quick and easy interim fix by loading the ghostery library as userscript in the background. I just looked into it a bit more and apparently it is not that easy (could also be my lack of knowledge though).

**Comment by alkim0** [comment]
I took a first stab at implementing cosmetic filtering and scriptlet injection.

There is a pending pull request: #7312 

To try it out for yourself:
1. I highly recommended adding ublock's filters.txt to your content.blocking.adblock.lists. Easylist doesn't have any scriptlet injection:
```
"https://github.com/uBlockOrigin/uAssets/raw/master/filters/filters.txt"
```
2. Update python-adblock to bleeding edge to get scriptlet injection:
```
pip install -U "git+https://github.com/alkim0/python-adblock.git"
```
3. Download qutebrowser from
```
git clone https://github.com/alkim0/qutebrowser
```
4. In the base directory, run `python qutebrowser.py`
5. In qutebrowser, run:
    1. `:adblock-update` (if you updated your adblock lists)
    2. `:adblock-update-resources`

At the very least, it should blocks youtube ads...

**Comment by ArniDagur** [comment]
That is crazy cool! I will review your pull request soon™

**Comment by alkim0** [comment]
> Thanks a LOT, do I understand correctly that there's still no dynamic scriptlet injection (just once after page load)? This could explain why ads where not completely blocked on some pages, such as distrowatch.org, heise.de and reddit.com (sponsored posts), when I just tried it out. It could also be that the corresponding blocking rules are on a different filter list (just used easylist and the default ublock list). Still, your work is a giant leap forward!

No, scriptlet injection should be working with the bleeding edge version of python-adblock (that's how we skip youtube ads).

I looked at the websites you mentioned. First, with heise.de, I get the same ads on firefox with ublock + umatrix, so those might not be blocked under the default filters.

Regarding distrowatch.org and reddit.com, I noticed that the underlying engine doesn't seem to be returning some of the more complex css filters. It was the case with these websites as well when I checked them out. I've filed it as an issue with python-adblock (https://github.com/ArniDagur/python-adblock/issues/71), but since that's a thin wrapper, I'm not sure the problem is there either...

**Comment by alkim0** [comment]
Wow, thanks for the detailed debug info.

Just to reiterate, css filters like `[id^="t3"].promotedlink:upward(.rpBJOHq2PR60pnwJlUyP0 > div)` (reddit's promoted posts) are not supported by the underlying Brave ad block engine (brave/adblock-rust/issues/145), so nothing can really be done about them at the moment.

**Comment by alkim0** [comment]
@herrsimon, Just to verify, is this what you are talking about?
![2022-07-15-105620_1916x1023_scrot](https://user-images.githubusercontent.com/9775404/179284550-c5bc7c4f-2231-489d-90f4-f3a73af5f11e.png)
If so, I still seem to be getting ads with it:
![2022-07-15-105502_1916x1023_scrot](https://user-images.githubusercontent.com/9775404/179284647-0a0ee1dd-5083-4f26-bea7-5bf9e7ba5e0f.png)

1. Qutebrowser doesn't seem to dealing with distrowatch.org's 301 properly at the moment, which is why the url doesn't change to distrowatch.com (which is why the adblocker won't pick up on it). I think this needs to be managed as a separate issue.

2. These are network requests and should be blocked by the existing adblocker already in qutebrowser. So, if there is a problem with this, that should be raised as a separate issue as well.

**Comment by Nicholas42** [comment]
> > As an aside, for people looking for a workaround for YouTube: Someone on Reddit [recently wrote a Greasemonkey script](https://old.reddit.com/r/qutebrowser/comments/ntl2ko/easy_youtube_adblocker_greasemonkey_script/h15yaqc?utm_source=share&utm_medium=web2x&context=3) to do so. I haven't tested it myself yet. Another approach is to use an external player like `mpv` instead, see 10. in the [FAQ](https://github.com/qutebrowser/qutebrowser/blob/master/doc/faq.asciidoc).
> 
> The Greasemonkey script is a great stopgap.
> 
> Here is a mirror in case something happens to the Reddit post. I have this in `~/.local/share/qutebrowser/greasemonkey/youtube-adblock.user.js`.
> 
> ```js
> // ==UserScript==
> // @name         Auto Skip YouTube Ads 
> // @version      1.0.1
> // @description  Speed up and skip YouTube ads automatically 
> // @author       jso8910
> // @match        *://*.youtube.com/*
> // @exclude      *://*.youtube.com/subscribe_embed?*
> // ==/UserScript==
> let main = new MutationObserver(() => {
>     let ad = [...document.querySelectorAll('.ad-showing')][0];
>     if (ad) {
>         let btn = document.querySelector('.videoAdUiSkipButton,.ytp-ad-skip-button')
>         if (btn) {
>             btn.click()
>         }
>     }
> })
> 
> main.observe(document.querySelector('.videoAdUiSkipButton,.ytp-ad-skip-button'), {attributes: true, characterData: true, childList: true})
> ```
> 
> If that doesn't work, try:
> 
> ```js
> // ==UserScript==
> // @name         Auto Skip YouTube Ads 
> // @version      1.0.0
> // @description  Speed up and skip YouTube ads automatically 
> // @author       jso8910
> // @match        *://*.youtube.com/*
> // @exclude      *://*.youtube.com/subscribe_embed?*
> // ==/UserScript==
> setInterval(() => {
>     const btn = document.querySelector('.videoAdUiSkipButton,.ytp-ad-skip-button')
>     if (btn) {
>         btn.click()
>     }
>     const ad = [...document.querySelectorAll('.ad-showing')][0];
>     if (ad) {
>         document.querySelector('video').playbackRate = 10;
>     }
> }, 50)
> ```

That works like a charm for me. I reworked it a bit to be faster on unskippable ads (those seem to be the most prevalent today). It seeks right to the end of the ad video, so you are done with them faster. I also tried to hide the video as much as possible:

```js
// ==UserScript==
// @name         Auto Skip YouTube Ads
// @version      1.1.0
// @description  Speed up and skip YouTube ads automatically
// @author       jso8910
// @match        *://*.youtube.com/*
// @exclude      *://*.youtube.com/subscribe_embed?*
// ==/UserScript==
setInterval(() => {
    const btn = document.querySelector('.videoAdUiSkipButton,.ytp-ad-skip-button')
    if (btn) {
        btn.click()
    }
    const ad = [...document.querySelectorAll('.ad-showing')][0];
    if (ad) {
        const video = document.querySelector('video')
        video.muted = true;
        video.hidden = true;

        // This is not necessarily available right at the start
        if(video.duration != NaN) {
            video.currentTime = video.duration;
        }

        // 16 seems to be the highest rate that works, mostly this isn't needed
        video.playbackRate = 16;
    }
}, 50)
```

**Comment by hakan-demirli** [comment]
I block everything I don't use on Youtube. For example Home, Shorts, ... buttons on sidebar, recommendations on the right etc. Since, the PR haven't merged I use greasemonkey scripts to block those elements.

Here is a simple script that removes "Home" button from from Youtube sidebar.
```js
// ==UserScript==
// @name         Remove Elements with Text (MutationObserver)
// @namespace    http://tampermonkey.net/
// @version      0.1
// @description  Remove elements within #items if they contain specified text(s)
// @author       You
// @match        *://*/*
// @grant        none
// ==/UserScript==

(function () {
  "use strict";

  // Function to remove elements within #items if they contain specified text
  function removeElementsWithText(texts) {
    var elementsToRemove = document.querySelectorAll(
      "#items > ytd-guide-entry-renderer",
    );

    Array.from(elementsToRemove).forEach(function (element) {
      texts.forEach(function (text) {
        if (element.innerText.includes(text)) {
          element.remove();
        }
      });
    });
  }

  let blist = ["Home", "Shorts"];

  // Function to be called when mutations are observed
  function handleMutations(mutationsList, observer) {
    removeElementsWithText(blist); // Adjust the texts as needed
  }

  // Options for the observer (which mutations to observe)
  const observerConfig = { childList: true, subtree: true };

  // Create an observer instance linked to the callback function
  const observer = new MutationObserver(handleMutations);

  // Start observing the target node for configured mutations
  observer.observe(document.body, observerConfig);

  // Perform initial removal on page load
  removeElementsWithText(blist); // Adjust the texts as needed
})();
```

To block any element just copy its JS path using `:devtools` -> right click -> `inspect` and then modify the script.

**Comment by simonjones49** [comment]
I just tried all of these and Youtube is just not filtered. It still shows sponsored ads in the video list and yes there is a script to fastforward the video but this is not adblocking. For me at least this alone was enough to return to Librewolf. I will check in to see if this can be resolved as I love the rest of the browser. Thank you for your work.
